File: composey/compiler/inference/_connectivity.py
"""Networking inference for AWS resources.

Handles security groups, network connectivity, and service discovery.
"""

from typing import Callable

from composey.models.aws import (
    AWSResources,
    SecurityGroup,
    SecurityGroupRule,
    ServiceDiscoveryPrivateDnsNamespace,
    ServiceDiscoveryService,
)
from composey.models.environment import AwsEnvironment
from composey.models.semantic import Application as SemanticApp
from composey.models.semantic import Service as SemanticService

from ._common import namespace_for, safe_terraform_identifier


def infer_networking(
    resources: AWSResources,
    app: SemanticApp,
    env: AwsEnvironment,
    get_name: Callable[[str], str],
    tags: dict[str, str] | None,
) -> None:
    """Infer security groups and network rules for the application.

    Creates one security group per compose network. Services sharing a network
    can reach each other, and services on disjoint networks cannot, which is
    what Compose already enforces locally.
    """
    networks = sorted({n for service in app.services for n in service.networks})

    for network in networks:
        key = _sg_key(network)
        resources.aws_security_group[key] = SecurityGroup(
            name=get_name(network),
            vpc_id=env.vpc_id,
            description=f"Network {network} of {app.name} in {env.name}",
            tags=tags,
        )

        # Members of a network reach each other on any port
        resources.aws_security_group_rule[f"{key}_internal_rule"] = SecurityGroupRule(
            type="ingress",
            from_port=0,
            to_port=0,
            protocol="-1",
            security_group_id=f"${{aws_security_group.{key}.id}}",
            source_security_group_id=f"${{aws_security_group.{key}.id}}",
            description=f"Allow traffic within network {network}",
        )

        # Allow outbound access (needed for pulling images, writing logs)
        resources.aws_security_group_rule[f"{key}_egress_rule"] = SecurityGroupRule(
            type="egress",
            from_port=0,
            to_port=0,
            protocol="-1",
            security_group_id=f"${{aws_security_group.{key}.id}}",
            cidr_blocks=["0.0.0.0/0"],
            description=f"Allow all outbound from network {network}",
        )


def infer_service_discovery(
    resources: AWSResources,
    app: SemanticApp,
    env: AwsEnvironment,
    get_name: Callable[[str], str],
    tags: dict[str, str] | None,
) -> str:
    """Create service discovery namespace and services.

    Returns the namespace name for use in connection strings.
    """
    namespace = namespace_for(env.name, app.name)
    discoverable = [s for s in app.services if _is_discoverable(s)]

    if discoverable:
        resources.aws_service_discovery_private_dns_namespace["app"] = (
            ServiceDiscoveryPrivateDnsNamespace(
                name=namespace,
                vpc=env.vpc_id,
                description=f"Service discovery for {app.name} in {env.name}",
                tags=tags,
            )
        )

    for service in discoverable:
        resources.aws_service_discovery_service[
            f"{safe_terraform_identifier(service.name)}_discovery"
        ] = ServiceDiscoveryService(
            name=service.name,
            dns_config={
                "namespace_id": "${aws_service_discovery_private_dns_namespace.app.id}",
                "dns_records": [{"ttl": 10, "type": "A"}],
                "routing_policy": "MULTIVALUE",
            },
            tags=tags,
        )

    return namespace


def security_group_ids(service_networks: list[str]) -> list[str]:
    """Get Terraform references to security groups for given networks."""
    return [
        f"${{aws_security_group.{_sg_key(n)}.id}}" for n in sorted(service_networks)
    ]


def _sg_key(network: str) -> str:
    """Generate a Terraform-safe security group key from a network name."""
    return f"{safe_terraform_identifier(network)}_sg"


def _is_discoverable(service: SemanticService) -> bool:
    """Check if a service can be discovered by other services.

    A scheduled task runs and exits rather than being a service, and a container
    with no port publishes nothing to reach.
    """
    return (
        service.capability == "container"
        and service.port is not None
        and service.schedule is None
    )

File: cloudcompose-go/internal/compiler/aws/connectivity.go
package aws

import (
	"fmt"
	"sort"

	"github.com/gecburton/cloudcompose/internal/models"
)

// sgKey generates a Terraform-safe security group key from a network name.
func sgKey(network string) string {
	return SafeTerraformIdentifier(network) + "_sg"
}

// SecurityGroupIDs gets Terraform references to security groups for given
// network segments.
func SecurityGroupIDs(serviceSegments []string) []string {
	sorted := make([]string, len(serviceSegments))
	copy(sorted, serviceSegments)
	sort.Strings(sorted)

	ids := make([]string, len(sorted))
	for i, n := range sorted {
		ids[i] = fmt.Sprintf("${aws_security_group.%s.id}", sgKey(n))
	}
	return ids
}

// IsDiscoverable checks if a service can be discovered by other services. A
// scheduled task runs and exits rather than being a service, and a container
// with no port publishes nothing to reach.
func IsDiscoverable(service *models.Service) bool {
	return service.Capability == models.CapabilityContainer &&
		service.Port != nil &&
		service.Schedule == nil
}

// InferNetworking infers security groups and network rules for the
// application. Creates one security group per network isolation segment.
// Services sharing a segment can reach each other, and services on disjoint
// segments cannot. This maps to AWS security groups; other clouds use
// equivalent mechanisms.
func InferNetworking(
	resources *models.AWSResources,
	app *models.Application,
	env *models.AwsEnvironment,
	getName func(string) string,
	tags map[string]string,
) {
	networkSet := map[string]struct{}{}
	for _, service := range app.Services {
		for _, n := range service.NetworkIsolationSegments {
			networkSet[n] = struct{}{}
		}
	}
	networks := make([]string, 0, len(networkSet))
	for n := range networkSet {
		networks = append(networks, n)
	}
	sort.Strings(networks)

	for _, network := range networks {
		key := sgKey(network)
		resources.SecurityGroup[key] = models.SecurityGroup{
			Name:        getName(network),
			VpcID:       env.VpcID,
			Description: fmt.Sprintf("Network %s of %s in %s", network, app.Name, env.Name),
			Tags:        tags,
		}

		selfRef := fmt.Sprintf("${aws_security_group.%s.id}", key)

		// Members of a network reach each other on any port.
		desc := fmt.Sprintf("Allow traffic within network %s", network)
		selfRefCopy := selfRef
		resources.SecurityGroupRule[key+"_internal_rule"] = models.SecurityGroupRule{
			Type:                  "ingress",
			FromPort:              0,
			ToPort:                0,
			Protocol:              "-1",
			SecurityGroupID:       selfRef,
			SourceSecurityGroupID: &selfRefCopy,
			Description:           &desc,
		}

		// Allow outbound access (needed for pulling images, writing logs).
		egressDesc := fmt.Sprintf("Allow all outbound from network %s", network)
		resources.SecurityGroupRule[key+"_egress_rule"] = models.SecurityGroupRule{
			Type:            "egress",
			FromPort:        0,
			ToPort:          0,
			Protocol:        "-1",
			SecurityGroupID: selfRef,
			CidrBlocks:      []string{"0.0.0.0/0"},
			Description:     &egressDesc,
		}
	}
}

// InferServiceDiscovery creates the service discovery namespace and
// services. Returns the namespace name for use in connection strings.
func InferServiceDiscovery(
	resources *models.AWSResources,
	app *models.Application,
	env *models.AwsEnvironment,
	getName func(string) string,
	tags map[string]string,
) string {
	namespace := NamespaceFor(env.Name, app.Name)

	var discoverable []models.Service
	for i := range app.Services {
		if IsDiscoverable(&app.Services[i]) {
			discoverable = append(discoverable, app.Services[i])
		}
	}

	if len(discoverable) > 0 {
		desc := fmt.Sprintf("Service discovery for %s in %s", app.Name, env.Name)
		resources.ServiceDiscoveryPrivateDnsNamespace["app"] = models.ServiceDiscoveryPrivateDnsNamespace{
			Name:        namespace,
			Vpc:         env.VpcID,
			Description: &desc,
			Tags:        tags,
		}
	}

	for _, service := range discoverable {
		svc := models.NewServiceDiscoveryService()
		svc.Name = service.Name
		svc.DnsConfig = map[string]any{
			"namespace_id":   "${aws_service_discovery_private_dns_namespace.app.id}",
			"dns_records":    []map[string]any{{"ttl": 10, "type": "A"}},
			"routing_policy": "MULTIVALUE",
		}
		svc.Tags = tags
		resources.ServiceDiscoveryService[SafeTerraformIdentifier(service.Name)+"_discovery"] = svc
	}

	return namespace
}
